runtime/lua/locals.lua

```diff
@@ -38,3 +41,102 @@ test[[
     return foo + bar
 ]]
 --> ~false\t.*attempt to reassign constant variable 'bar'
+
+--
+-- to-be-closed tests
+--
+
+test[[
+    local x <close> = nil
+    x = 3
+]]
+--> ~false\t.*attempt to reassign constant variable 'x'
+
+test[[
+    local x <close> = 1
+]]
+--> ~false\t.*to be closed variable missing a __close metamethod
+
+function make(msg)
+    t = {}
+    setmetatable(t, {__close = function () print(msg) end})
+    return t
+end
+
+do
+    local x <close> = make("x")
+    print("a")
+end
+print("b")
+--> =a
+--> =x
+--> =b
+
+do
+    local x <close> = make("x")
+    local y <close> = make("y")
+end
+--> =y
+--> =x
+
+do
+    local x <close>, y <close> = make("aa"), make("bb")
+end
+--> =bb
+--> =aa
+
+
+-- A function to test to-be-closed variables
+local s
+function mk(a)
+    t = {}
+    s = s .. '+' .. a
+    setmetatable(t, {__close = function () s = s .. '-' .. a end})
+    return t
+end
+
+-- How it works
+do
+    s = "start"
+    local v <close> = mk("bob")
+    print(s)
+    --> =start+bob
+end
+print(s)
+--> =start+bob-bob
+
+do
+    s = "start"
+    local function f()
+        local a <close> = mk("a")
+        for i = 1, 3 do
+            local b <close> = mk("b"..i)
+        end
+        do
+            local c <close> = mk("c")
+            do
+                local d <close> = mk("d")
+                return
+            end
+        end
+    end
+    f()
+    print(s)
+    --> =start+a+b1-b1+b2-b2+b3-b3+c+d-d-c-a
+end
+
+do
+    s = "start"
+    local function f(n)
+        local x <close> = mk("x"..n)
+        if n > 0 then
+            f(n - 1)
+        else
+            error("stop")
+        end
+    end
+    print(pcall(f, 3))
+    --> ~false\t.*: stop
+    print(s)
+    --> =start+x3+x2+x1+x0-x0-x1-x2-x3
+end
```

Comment:
We should probably also add a test for a failing `_close` function, like:
```go
__close = function () print(free_var) end
```

Reply:
Yeah, will do - I was unsure of the semantics initially so didn't write tests for that kind of stuff.